### PR TITLE
fix(routing): actions should redirect the original pathname

### PR DESCRIPTION
**Pull request by ematipico**
## Changes

Closes https://github.com/withastro/astro/issues/12146
Closes PLT-2576

An Astro action uses `context.url.pathname`  for its redirect, however this is a bug because when there's a _rewrite_, the `APIContext` gets updated **in place** when using `next(/*payload*/). Astro actions need to redirect to the same original URL where the initial request started.

To fix the bug, we inject a well known header when we create `RenderContext`, so no matter how many rewrites happen, the original pathname is always there and the same

## Testing

Created a new e2e test

## Docs

N/A

### CODE REVIEW

**Review by matthewp — APPROVED** [review]

**Review by bluwy — APPROVED** [review]

**Comment by bluwy** [comment]
Didn't notice the tests are failing 🙈 Could be related to copying the request

**Comment by ematipico** [comment]
> Didn't notice the tests are failing 🙈 Could be related to copying the request

Most probably, yeah. I think it's fine *to not* **try** to copy the body, and copy the request normally. Let's see if that works

File: packages/astro/e2e/actions-blog.test.js [hunk 1]
```diff
@@ -135,4 +135,17 @@ test.describe('Astro Actions - Blog', () => {
 		await logoutButton.click();
 		await expect(page).toHaveURL(astro.resolveUrl('/blog/'));
 	});
+
+	test('Should redirect to the origin pathname when there is a rewrite', async ({
+		page,
+		astro,
+	}) => {
+		await page.goto(astro.resolveUrl('/sum'));
+		//
```
Comment (bluwy) [inline packages/astro/e2e/actions-blog.test.js:144]:
```suggestion
		await page.goto(astro.resolveUrl('/sum'));
```

we could remove this stray comment i think

**Review by bluwy — APPROVED** [review]
Test fix looks great! The remaining fail seems expected from main

**Comment by sergeysova** [comment]
Looks like I have similar crash in production:

```
[info]TypeError: immutable
[info]    at appendHeader (node:internal/deps/undici/undici:8550:15)
[info]    at _Headers.append (node:internal/deps/undici/undici:8752:16)
[info]    at NodeApp.render (file:///class/dist/server/chunks/_@astrojs-ssr-adapter_DTyDJ9ui.mjs:870:26)
[info]    at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
[info]    at async file:///class/dist/server/chunks/_@astrojs-ssr-adapter_DTyDJ9ui.mjs:1204:30
```

astro: 4.16.7

**Comment by sergeysova** [comment]
Downgrade to 4.16.6 fixed

**Pull request merged**